File: Serene/Serene.Script/Common/Helper/BulkServiceAction.cs
/* 
This class has been ported to TypeScript. See OrderGrid.ts
Code below is only a reference for those who want to use Saltaralle

namespace Serene
{
    using Serenity;
    using System;
    using System.Collections.Generic;
    using System.Html;
    using System.Linq;
    using System.Runtime.CompilerServices;

    public abstract class BulkServiceAction : ScriptContext
    {
        protected List<string> keys;
        protected IEnumerator<string> queue;
        protected BasicProgressDialog progressDialog;
        protected int pendingRequests;
        protected int completedRequests;
        private int successCount;
        private int errorCount;
        protected Dictionary<string, ServiceError> errorByKey;

        protected virtual void CreateProgressDialog()
        {
            progressDialog = new BasicProgressDialog();
            progressDialog.DialogOpen();
            progressDialog.Max = keys.Count();
            progressDialog.Value = 0;
        }

        protected virtual string GetConfirmationFormat()
        {
            return Q.Text("Site.BulkServiceAction.ConfirmationFormat");
        }

        protected virtual string GetConfirmationMessage(int targetCount)
        {
            return String.Format(GetConfirmationFormat(), targetCount);
        }

        protected virtual void Confirm(int targetCount, Action action)
        {
            Q.Confirm(GetConfirmationMessage(targetCount), action);
        }

        protected virtual string GetNothingToProcessMessage()
        {
            return Q.Text("Site.BulkServiceAction.NothingToProcess");
        }

        protected virtual void NothingToProcess()
        {
            Q.NotifyError(GetNothingToProcessMessage());
        }

        protected virtual int GetParallelRequests()
        {
            return 1;
        }

        protected virtual int GetBatchSize()
        {
            return 1;
        }

        protected virtual void StartParallelExecution()
        {
            CreateProgressDialog();

            this.pendingRequests = 0;
            this.completedRequests = 0;
            this.errorCount = 0;
            this.errorByKey = new Dictionary<string, ServiceError>();
            this.queue = keys.GetEnumerator();

            int parallelRequests = GetParallelRequests();
            while (parallelRequests-- > 0)
                ExecuteNextBatch();
        }

        protected virtual void ServiceCallCleanup()
        {
            pendingRequests--;
            completedRequests++;

            string title = Q.Text(progressDialog.Cancelled ?
                "Site.BasicProgressDialog.CancelTitle" :
                "Site.BasicProgressDialog.PleaseWait");

            title += " (";

            if (successCount > 0)
                title += String.Format(Q.Text("Site.BulkServiceAction.SuccessCount"), successCount);

            if (errorCount > 0)
            {
                if (successCount > 0)
                    title += ", ";

                title += String.Format(Q.Text("Site.BulkServiceAction.ErrorCount"), errorCount);
            }

            progressDialog.Title = title + ")";
            progressDialog.Value = successCount + errorCount;

            if (!progressDialog.Cancelled && progressDialog.Value < keys.Count)
                ExecuteNextBatch();
            else if (pendingRequests == 0)
            {
                progressDialog.DialogClose();
                ShowResults();
                if (Done != null)
                {
                    Done();
                    Done = null;
                }
            }
        }

        protected virtual void ExecuteForBatch(List<string> batch)
        {
        }

        protected virtual void ExecuteNextBatch()
        {
            var batchSize = GetBatchSize();
            var batch = new List<string>();
            while (true)
            {
                if (batch.Count >= batchSize)
                    break;

                if (!queue.MoveNext())
                    break;

                batch.Add(queue.Current);
            }

            if (batch.Count > 0)
            {
                pendingRequests++;
                ExecuteForBatch(batch);
            }
        }

        protected virtual string GetAllHadErrorsFormat()
        {
            return Q.Text("Site.BulkServiceAction.AllHadErrorsFormat");
        }

        protected virtual void ShowAllHadErrors()
        {
            Q.NotifyError(String.Format(GetAllHadErrorsFormat(), errorCount));
        }

        protected virtual string GetSomeHadErrorsFormat()
        {
            return Q.Text("Site.BulkServiceAction.SomeHadErrorsFormat");
        }

        protected virtual void ShowSomeHadErrors()
        {
            Q.NotifyWarning(String.Format(GetSomeHadErrorsFormat(), successCount, errorCount));
        }

        protected virtual string GetAllSuccessFormat()
        {
            return Q.Text("Site.BulkServiceAction.AllSuccessFormat");
        }

        protected virtual void ShowAllSuccess()
        {
            Q.NotifySuccess(String.Format(GetAllSuccessFormat(), successCount));
        }

        protected virtual void ShowResults()
        {
            if (errorCount == 0 &&
                successCount == 0)
            {
                NothingToProcess();
                return;
            }

            if (errorCount > 0 &&
                successCount == 0)
            {
                ShowAllHadErrors();
                return;
            }

            if (errorCount > 0)
            {
                ShowSomeHadErrors();
                return;
            }

            ShowAllSuccess();
        }

        public virtual void Execute(List<string> keys)
        {
            this.keys = keys;

            if (this.keys.Count == 0)
            {
                NothingToProcess();
                return;
            }

            Confirm(this.keys.Count, () =>
            {
                StartParallelExecution();
            });
        }

        protected int SuccessCount
        {
            get { return successCount; }
            set { successCount = value; }
        }

        protected int ErrorCount
        {
            get { return errorCount; }
            set { errorCount = value; }
        }

        public Action Done;
    }
}
*/

File: Serene/Serene.Web/Modules/Common/Helpers/BulkServiceAction.ts
namespace Serene.Common {

    export class BulkServiceAction {

        protected keys: string[];
        protected queue: string[];
        protected queueIndex: number;
        protected progressDialog: BasicProgressDialog;
        protected pendingRequests: number;
        protected completedRequests: number;
        protected errorByKey: Q.Dictionary<Serenity.ServiceError>;
        private successCount;
        private errorCount;
        public done: () => void;

        protected createProgressDialog() {
            this.progressDialog = new BasicProgressDialog();
            this.progressDialog.dialogOpen();
            this.progressDialog.max = this.keys.length;
            this.progressDialog.value = 0;
        }

        protected getConfirmationFormat() {
            return Q.text('Site.BulkServiceAction.ConfirmationFormat');
        }

        protected getConfirmationMessage(targetCount) {
            return Q.format(this.getConfirmationFormat(), targetCount);
        }

        protected confirm(targetCount, action) {
            Q.confirm(this.getConfirmationMessage(targetCount), action);
        }

        protected getNothingToProcessMessage() {
            return Q.text('Site.BulkServiceAction.NothingToProcess');
        }

        protected nothingToProcess() {
            Q.notifyError(this.getNothingToProcessMessage());
        }

        protected getParallelRequests() {
            return 1;
        }

        protected getBatchSize() {
            return 1;
        }

        protected startParallelExecution() {
            this.createProgressDialog();
            this.successCount = 0;
            this.errorCount = 0;
            this.pendingRequests = 0;
            this.completedRequests = 0;
            this.errorCount = 0;
            this.errorByKey = {};
            this.queue = this.keys.slice();
            this.queueIndex = 0;
            var parallelRequests = this.getParallelRequests();
            while (parallelRequests-- > 0) {
                this.executeNextBatch();
            }
        }

        protected serviceCallCleanup() {
            this.pendingRequests--;
            this.completedRequests++;

            var title = Q.text((this.progressDialog.cancelled ?
                'Site.BasicProgressDialog.CancelTitle' : 'Site.BasicProgressDialog.PleaseWait'));

            title += ' (';
            if (this.successCount > 0) {
                title += Q.format(Q.text('Site.BulkServiceAction.SuccessCount'), this.successCount);
            }

            if (this.errorCount > 0) {
                if (this.successCount > 0) {
                    title += ', ';
                }

                title += Q.format(Q.text('Site.BulkServiceAction.ErrorCount'), this.errorCount);
            }

            this.progressDialog.title = title + ')';
            this.progressDialog.value = this.successCount + this.errorCount;
            if (!this.progressDialog.cancelled && this.progressDialog.value < this.keys.length) {
                this.executeNextBatch();
            }

            else if (this.pendingRequests === 0) {
                this.progressDialog.dialogClose();
                this.showResults();
                if (this.done) {
                    this.done();
                    this.done = null;
                }
            }
        }

        protected executeForBatch(batch: string[]) {
        }

        protected executeNextBatch() {
            var batchSize = this.getBatchSize();
            var batch = [];
            while (true) {
                if (batch.length >= batchSize) {
                    break;
                }

                if (this.queueIndex >= this.queue.length) {
                    break;
                }

                batch.push(this.queue[this.queueIndex++]);
            }

            if (batch.length > 0) {
                this.pendingRequests++;
                this.executeForBatch(batch);
            }
        }

        protected getAllHadErrorsFormat() {
            return Q.text('Site.BulkServiceAction.AllHadErrorsFormat');
        }

        protected showAllHadErrors() {
            Q.notifyError(Q.format(this.getAllHadErrorsFormat(), this.errorCount));
        }

        protected getSomeHadErrorsFormat() {
            return Q.text('Site.BulkServiceAction.SomeHadErrorsFormat');
        }

        protected showSomeHadErrors() {
            Q.notifyWarning(Q.format(this.getSomeHadErrorsFormat(), this.successCount, this.errorCount));
        }

        protected getAllSuccessFormat() {
            return Q.text('Site.BulkServiceAction.AllSuccessFormat');
        }

        protected showAllSuccess() {
            Q.notifySuccess(Q.format(this.getAllSuccessFormat(), this.successCount));
        }

        protected showResults() {
            if (this.errorCount === 0 && this.successCount === 0) {
                this.nothingToProcess();
                return;
            }

            if (this.errorCount > 0 && this.successCount === 0) {
                this.showAllHadErrors();
                return;
            }

            if (this.errorCount > 0) {
                this.showSomeHadErrors();
                return;
            }

            this.showAllSuccess();
        }

        public execute(keys: string[]) {
            this.keys = keys;
            if (this.keys.length === 0) {
                this.nothingToProcess();
                return;
            }
            this.confirm(this.keys.length, () => this.startParallelExecution());
        }

        get_successCount() {
            return this.successCount;
        }

        set_successCount(value: number) {
            this.successCount = value;
        }

        get_errorCount() {
            return this.errorCount;
        }

        set_errorCount(value: number) {
            this.errorCount = value;
        }
    }
}